### PR TITLE
Add extension to handle NSTextAttachment as KingfisherCompatible

### PR DESCRIPTION
**Pull request by PalringoBen**
This will allow using Kingfisher on NSTextAttachment in an attributed label. I've followed the same format for other extensions that add compliance for KingfisherCompatible.

We can now add a image to a label using Kingfishers cache like this
```
// Create a label
let label = UILabel()

// Create a attributed string
let attributedText = NSMutableAttributedString(string: "some string")

// Create the text attachment
let textAttachment = NSTextAttachment()

// Set the size of the attachment
let capHeight = label?.font.capHeight ?? 0
textAttachment.bounds = CGRect(x: 0, y: -1, width: capHeight + 2, height: capHeight + 2)

// Request the image from Kingfisher
textAttachment.kf.setImage(with: url)

// Add the text attachment to the beging of the text attributedText
attributedText.replaceCharacters(in: NSRange(), with: NSAttributedString(attachment: textAttachment))

// set the attributedText on the label
label.attributedText = attributedText
```

### CODE REVIEW

**Comment by onevcat** [comment]
It looks nice! Thanks.

I will take another look at it soon and check whether we could have a better way for creating extension for other UI components.

**Comment by onevcat** [comment]
@PalringoBen Terribly sorry for the delay!

I tried the related changes, but it seems the extension code only works fine when loading the image from memory.

For images loaded from network or disk (both are in an asynchronous way), it shows a file icon instead and never changed to the actual image:

Snip20190809_1

I can make it work by setting the `attributedText` in the completion handler in the `kf.setImage` mehtod:

```swift
textAttachment.kf.setImage(with: url) { r in
    attributedText.replaceCharacters(in: NSRange(), with: NSAttributedString(attachment: textAttachment))
    label.attributedText = attributedText
}
```

or at least, I need to call `setNeedsDisplay()` in the handler:

```swift
textAttachment.kf.setImage(with: url) { r in
    label.setNeedsDisplay()
}

attributedText.replaceCharacters(in: NSRange(), with: NSAttributedString(attachment: textAttachment))
label.attributedText = attributedText
```

Snip20190809_2

I am not sure whether it is what you meant to do or did you notice this? In the ideal world, the label refreshing is expected to happen automatically when we set the attachment image property (I know UIKit will not do that but for easier using, maybe we should find a way for that). 

Do you have any idea about it? Or am I wrong on something?

File: Sources/Extensions/NSTextAttachment+Kingfisher.swift [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                    switch result {
+                    case .success(let value):
+                        self.base.image = value.image
+                        completionHandler?(result)
+
+                    case .failure:
+                        if let image = options.onFailureImage {
+                            self.base.image = image
+                        }
+                        completionHandler?(result)
```
Comment (Adobels) [inline Sources/Extensions/NSTextAttachment+Kingfisher.swift]: Multiple cases call the same `completionHandler` method. This call should be placed after the switch statement.

File: Sources/Extensions/NSTextAttachment+Kingfisher.swift [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        let issuedIdentifier = Source.Identifier.next()
+        mutatingSelf.taskIdentifier = issuedIdentifier
+
+        if let block = progressBlock {
+            options.onDataReceived = (options.onDataReceived ?? []) + [ImageLoadingProgressSideEffect(block)]
+        }
+
+        if let provider = ImageProgressiveProvider(options, refresh: { image in
```
Comment (emrcftci) [inline Sources/Extensions/NSTextAttachment+Kingfisher.swift:83]: You can use like this
```swift
ImageProgressiveProvider(options) { image in
    self.base.image = image
} { ...
```

File: Sources/Extensions/NSTextAttachment+Kingfisher.swift [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    /// Nothing will happen if the downloading has already finished.
+    public func cancelDownloadTask() {
+        imageTask?.cancel()
+    }
+}
+
+private var taskIdentifierKey: Void?
+private var imageTaskKey: Void?
+
+// MARK: Properties
```
Comment (emrcftci) [inline Sources/Extensions/NSTextAttachment+Kingfisher.swift:178]: please leave a whitespace after `// MARK`

**Comment by CLoutas** [comment]
Has this been abandoned? Is there another solution to handle `NSTextAttachment`?

**Comment by nootfly** [comment]
Any update for this issue?

**Comment by onevcat** [comment]
Merged as a part of #1495 

I will tag a new version for it soon.

**Comment by espenbye** [comment]
Hi @onevcat! This broke Kingfisher when targeting watchOS, as NSTextAttachment isn't available there.

> /SourcePackages/checkouts/Kingfisher/Sources/Extensions/NSTextAttachment+Kingfisher.swift:34:41: Cannot find type 'NSTextAttachment' in scope

**Comment by onevcat** [comment]
watchOS build should be fixed in #1499

**Comment by fukemy** [comment]
hello, i updated pod to Installing Kingfisher (5.14.1) but can not find NSTextAttachment+Kingfisher, can u help?

**Comment by fukemy** [comment]
Screen Shot 2020-10-16 at 16 25 04
here is error

**Comment by onevcat** [comment]
@fukemy It is a part of 5.15.0.

**Pull request merged**